### ISSUE TITLE
Remove bundled flatbuffers module

### ISSUE THREAD

**Issue by herbygillot** [crossbario/zlmdb#28]
* ZLMDB version: 20.8.1
* Python version: 3.8.5
* Operating System: macOS 10.15.6 (Catalina)

### Description
Hello, it seems to be that `zlmdb` is exporting the `flatbuffers` package.

I say this because while packaging `zlmdb`, I see it installing a `flatbuffers` directory into place in `.../site-packages`, and also per this line here in 20.8.1's `setup.py`:

https://github.com/crossbario/zlmdb/blob/1eabd29c4a4663a8d78468f6cca343615fca5676/setup.py#L62

This is causing issues when trying to install the official `flatbuffers` package separately.

I'm just trying to understand if this is indeed the case, and if so, why it would be necessary?

**Comment by oberstet** [comment]
The original reason was:

* the official flatbuffers package on PyPI lagged behind the latest version on GitHub - I seem to remember that I wanted to have https://github.com/google/flatbuffers/pull/4713
* we hence added flatbuffers here, and exported it too, so we can access the latest module

however, I can see how this leads to problems .. the installation as a top-level module in site-packages.

I can see 2 solutions:

a) remove the bundled module and rely on upstream PyPI 
b) bundle and install flatbuffers as a submodule

**Comment by herbygillot** [comment]
OK understood.  I would like to take this opportunity to put __all my weight__ behind a vote for **a)**, which is to *remove* the bundled module and *rely on upstream PyPI*.  Option a) makes `zlmdb` play nice especially in environments where one may not have the luxury of `virtualenv`, and it also makes `zlmdb` much easier for maintainers to package and deploy.

Thank you for your insight here @oberstet.

**Comment by oberstet** [comment]
ok, just checked again: so the PyPI distributed package does not contain the reflection schema (`reflection.fbs`) nor the python wrapper classes generated from that. in my eyes, those should be part of the FB upstream package, but since they currently aren't:

```
(cpy382_1) oberstet@intel-nuci7:~/scm/crossbario/zlmdb$ ll flatbuffers/
insgesamt 68
drwxr-xr-x 2 oberstet oberstet  4096 Aug 17 12:18 ./
drwxr-xr-x 9 oberstet oberstet  4096 Aug 17 12:18 ../
-rw-r--r-- 1 oberstet oberstet 25217 Aug 14 23:45 builder.py
-rw-r--r-- 1 oberstet oberstet  2342 Aug 14 23:45 compat.py
-rw-r--r-- 1 oberstet oberstet  1555 Aug 14 23:45 encode.py
-rw-r--r-- 1 oberstet oberstet   698 Aug 14 23:45 __init__.py
-rw-r--r-- 1 oberstet oberstet  3955 Aug 14 23:45 number_types.py
-rw-r--r-- 1 oberstet oberstet  1165 Aug 14 23:45 packer.py
-rw-r--r-- 1 oberstet oberstet  4708 Aug 14 23:45 table.py
-rw-r--r-- 1 oberstet oberstet   994 Aug 14 23:45 util.py
(cpy382_1) oberstet@intel-nuci7:~/scm/crossbario/zlmdb$ tar tvzf ~/flatbuffers-1.12.tar.gz
drwxr-x--- wvo/primarygroup  0 2020-03-13 00:32 flatbuffers-1.12/
-rw-r--r-- wvo/primarygroup 67 2020-03-13 00:32 flatbuffers-1.12/setup.cfg
-rw-r----- wvo/primarygroup 882 2020-03-13 00:32 flatbuffers-1.12/PKG-INFO
-rw-r--r-- wvo/primarygroup 2955 2019-05-31 20:29 flatbuffers-1.12/setup.py
drwxr-x--- wvo/primarygroup    0 2020-03-13 00:32 flatbuffers-1.12/flatbuffers.egg-info/
-rw-r--r-- wvo/primarygroup   12 2020-03-13 00:32 flatbuffers-1.12/flatbuffers.egg-info/top_level.txt
-rw-r--r-- wvo/primarygroup  340 2020-03-13 00:32 flatbuffers-1.12/flatbuffers.egg-info/SOURCES.txt
-rw-r--r-- wvo/primarygroup  882 2020-03-13 00:32 flatbuffers-1.12/flatbuffers.egg-info/PKG-INFO
-rw-r--r-- wvo/primarygroup    1 2020-03-13 00:32 flatbuffers-1.12/flatbuffers.egg-info/dependency_links.txt
drwxr-x--- wvo/primarygroup    0 2020-03-13 00:32 flatbuffers-1.12/flatbuffers/
-rw-r----- wvo/primarygroup 2342 2018-02-08 23:51 flatbuffers-1.12/flatbuffers/compat.py
-rw-r----- wvo/primarygroup 1165 2018-02-08 23:51 flatbuffers-1.12/flatbuffers/packer.py
-rw-r--r-- wvo/primarygroup 1581 2019-07-27 02:49 flatbuffers-1.12/flatbuffers/encode.py
-rw-r----- wvo/primarygroup 4708 2018-02-08 23:51 flatbuffers-1.12/flatbuffers/table.py
-rw-r--r-- wvo/primarygroup 25858 2020-03-10 00:46 flatbuffers-1.12/flatbuffers/builder.py
-rw-r----- wvo/primarygroup  3955 2018-02-08 23:51 flatbuffers-1.12/flatbuffers/number_types.py
-rw-r--r-- wvo/primarygroup  1669 2019-07-27 02:49 flatbuffers-1.12/flatbuffers/util.py
-rw-r--r-- wvo/primarygroup   717 2020-03-02 21:16 flatbuffers-1.12/flatbuffers/__init__.py
(cpy382_1) oberstet@intel-nuci7:~/scm/crossbario/zlmdb$ ll zlmdb/flatbuffers/
insgesamt 36
drwxr-xr-x 4 oberstet oberstet 4096 Aug 17 12:18 ./
drwxr-xr-x 4 oberstet oberstet 4096 Aug 17 12:18 ../
drwxr-xr-x 4 oberstet oberstet 4096 Aug 17 12:18 demo/
-rw-r--r-- 1 oberstet oberstet  441 Aug 14 23:45 demo1.fbs
-rw-r--r-- 1 oberstet oberstet 5322 Aug 14 23:45 demo2.fbs
-rw-r--r-- 1 oberstet oberstet 1293 Aug 14 23:45 __init__.py
drwxr-xr-x 2 oberstet oberstet 4096 Aug 14 23:45 reflection/
lrwxrwxrwx 1 oberstet oberstet   48 Aug 14 23:45 reflection.fbs -> ../../deps/flatbuffers/reflection/reflection.fbs
-rw-r--r-- 1 oberstet oberstet   90 Aug 14 23:45 zlmdb.fbs
(cpy382_1) oberstet@intel-nuci7:~/scm/crossbario/zlmdb$ 
```

> another issue is the lack of proper module version: https://github.com/google/flatbuffers/issues/5299

---

that being said, we should be able to remove https://github.com/crossbario/zlmdb/tree/master/flatbuffers ..

**Comment by oberstet** [comment]
will land via https://github.com/crossbario/zlmdb/pull/29 ..

**Comment by oberstet** [comment]
no, upstream is again lagging: https://github.com/google/flatbuffers/tree/master/python/flatbuffers (eg the `_version.py` file is missing from the published package ..)

we'll continue to use our own bundled version for now. pls bug upstream to rectify/update their py release process

**Comment by ryandesign** [comment]
I would not say that "upstream is ... lagging"; it merely appears that their latest release was in March, and the `_version.py` change was made after that. They should release a new version.

**Comment by ryandesign** [comment]
> the PyPI distributed package does not contain the reflection schema (`reflection.fbs`) nor the python wrapper classes generated from that. in my eyes, those should be part of the FB upstream package

And is this something that their current master does, so we are just waiting on a new release? Or is this something their current master does not do, in which case has an issue been filed with them about this?

**Comment by oberstet** [comment]
re: "lagging": yes, upstream did not yet publish a release containing https://github.com/google/flatbuffers/issues/5299

once that is releases, we can complete https://github.com/crossbario/zlmdb/pull/29

rgd `reflection.fbs` and the Python classes generated: as mentioned, I think these should also be bundled upstream - which they are not currently (I didn't bother to file an issue or PR though, my bad ..). however, that's yet another thing .. we could complete #29 without that

I'll keep this issue and PR open for now ..

**Comment by ryandesign** [comment]
> I would like to take this opportunity to put **all my weight** behind a vote for **a)**, which is to _remove_ the bundled module and _rely on upstream PyPI_.

So after further consultation with the developers of flatbuffers, specifically https://github.com/google/flatbuffers/issues/6076#issuecomment-694954041, it seems that we should go the other way. zlmdb and all other projects that use flatbuffers should continue to bundle flatbuffers, but in a way that installs any flatbuffers files in a non-conflicting way, and MacPorts and other package managers should not provide standalone flatbuffers for any language.

**Comment by oberstet** [comment]
> but in a way that installs any flatbuffers files in a non-conflicting way, and MacPorts and other package managers should not provide standalone flatbuffers for any language.

yes: just use a venv.

other than that: would be tricky, because we would need to move the module to a different name (eg `zlmdb.flatbuffers` here https://github.com/crossbario/zlmdb/blob/6e153b19f60377de9425fc09d5ffd31ca6bee0ad/setup.py#L62), but the Python code generated by flatc assumes a fixed name `flatbuffers`. so flatc would need an option to _set_ the name of the module it assumes to import that helper code ...

IOW: since that option is not available in flatc today, we would need upstream to do that here in zlmdb first ..

**Comment by ryandesign** [comment]
> yes: just use a venv.

It's nice that the virtualenv feature exists for those who want to use that, but I don't. I want to manage the software I install using MacPorts, and MacPorts does not use virtualenv to install its Python modules, and nobody has ever proposed any plan to change that. So I want to be able to install crossbar, thus zlmdb and its flatbuffers, "normally", not using virtualenv.

**Comment by oberstet** [comment]
won't fix (see above)
